src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4979,5 +4794,1571 @@ await foreach (var o in new Enumerable())
+            comp.VerifyDiagnostics(
+                // (8,33): error CS0121: The call is ambiguous between the following methods or properties: 'Extensions1.GetAsyncEnumerator(C)' and 'Extensions2.GetAsyncEnumerator(C)'
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.ERR_AmbigCall, "new C()").WithArguments("Extensions1.GetAsyncEnumerator(C)", "Extensions2.GetAsyncEnumerator(C)").WithLocation(8, 33)
+                );
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaAmbiguousExtensionsWhenOneHasCorrectPattern()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}
+public static class Extensions1
+{
+    public static int GetAsyncEnumerator(this C self) => 42;
+}
+public static class Extensions2
+{
+    public static C.Enumerator GetAsyncEnumerator(this C self) => new C.Enumerator();
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (8,33): error CS0121: The call is ambiguous between the following methods or properties: 'Extensions1.GetAsyncEnumerator(C)' and 'Extensions2.GetAsyncEnumerator(C)'
+                    //         await foreach (var i in new C())
+                    Diagnostic(ErrorCode.ERR_AmbigCall, "new C()").WithArguments("Extensions1.GetAsyncEnumerator(C)", "Extensions2.GetAsyncEnumerator(C)").WithLocation(8, 33)
+                    );
+        }
+
+        [Fact]
+        public void TestGetAsyncEnumeratorPatternViaAmbiguousExtensionsWhenNeitherHasCorrectPattern()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+            Console.Write(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public Task<bool> MoveNextAsync() => Task.FromResult(Current++ != 3);
+    }
+}
+public static class Extensions1
+{
+    public static int GetAsyncEnumerator(this C self) => 42;
+}
+public static class Extensions2
+{
+    public static bool GetAsyncEnumerator(this C self) => true;
+}";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (8,33): error CS0121: The call is ambiguous between the following methods or properties: 'Extensions1.GetAsyncEnumerator(C)' and 'Extensions2.GetAsyncEnumerator(C)'
```

Comment:
No, it's not, neither of these are applicable. I would not expect this error here.

Reply:
I guess for this one, there's not much we can do, because this is result of overload resolution, before the methods would be thrown out due to their return types.

---
In reply to: [412367333](https://github.com/dotnet/roslyn/pull/43050#discussion_r412367333) [](ancestors = 412367333)

Reply:
Exactly